[Huang Jiaxi] iP

huangjiaxi1111 opened the pull request:
# Killua
> “Assassination — It's the family trade. We all take it up. My folks see me as an exceptional prospect. But I don't see that I should have to live up to their expectations.” – Killua Zoldyck ([source](https://hero.fandom.com/wiki/Killua_Zoldyck))

Killua frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/huangjiaxi1111/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 👍 

And it is **FREE!**

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello!");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

huangjiaxi1111 committed:
greet and exit

huangjiaxi1111 committed:
echo

huangjiaxi1111 committed:
add, list

huangjiaxi1111 committed:
mark as done

huangjiaxi1111 committed:
todos, events, deadlines

huangjiaxi1111 committed:
automated text ui testing

huangjiaxi1111 committed:
handle errors

huangjiaxi1111 committed:
handle errors

huangjiaxi1111 committed:
delete

huangjiaxi1111 committed:
use enums

huangjiaxi1111 committed:
added user guide

huangjiaxi1111 committed:
Fix styling issues

huangjiaxi1111 committed:
Add saveList to save task history in ./data/tasklist.txt

huangjiaxi1111 committed:
Add loadTasks and parseTask to load task history when chatbot starts up

huangjiaxi1111 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add loadTasks and parseTask to load task history when chatbot starts up
  Add saveList to save task history in ./data/tasklist.txt

huangjiaxi1111 committed:
Fix bug: handle the case where the data file doesn't exist at the start.

huangjiaxi1111 committed:
Add dates and times; Add additional command "on"

Deadline and event now stores LocalDate or LocalDateTime instead of strings.
Command "on" takes in LocalDate and output tasks on that day.

huangjiaxi1111 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add dates and times; Add additional command "on"

huangjiaxi1111 committed:
Implement more OOP

Add Ui class to handle interactions with user
Add Storage class to store and load task history
Add Parser class to parse command
Add Tasklist class to contain the task list
Add abstract Command class and subclasses to execute command

huangjiaxi1111 committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  Implement more OOP

huangjiaxi1111 committed:
Divide classes into packages

huangjiaxi1111 committed:
Merge branch 'A-Packages'

* A-Packages:
  Divide classes into packages

huangjiaxi1111 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

huangjiaxi1111 committed:
Fix gradle.build; Fix bug with exception

Set mainClass to correct class. Make invalid-command-exception caught before argument-empty-exception

huangjiaxi1111 committed:
Add some JUnit tests

Test Deadline, Event, Todo, Parser, and KilluaException

huangjiaxi1111 committed:
Merge branch 'A-JUnit'

* A-JUnit:
  Add some JUnit tests

huangjiaxi1111 committed:
Change welcome page

huangjiaxi1111 committed:
Add JavaDoc

huangjiaxi1111 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc

huangjiaxi1111 committed:
Fix code style

huangjiaxi1111 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fix code style

huangjiaxi1111 committed:
Add find command to find a task by searching for a keyword

huangjiaxi1111 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find command to find a task by searching for a keyword

# Conflicts:
#	src/main/java/killua/task/Task.java

Resolution: adjust position of newly added method and its javadoc

huangjiaxi1111 committed:
Fix coding style violation

import wildcard is removed
Ui, Storage, Parser moved under package ui, storage, parser

hovanleong commented on src/main/java/killua/task/Task.java:
```diff
@@ -0,0 +1,86 @@
+ * Represents a generic task in the Killua application.
+ * This class provides basic functionalities for any type of task, including description and completion status.
+ */
+public class Task {
```
Would you be open to considering making Task an abstract class, since you are not likely to have to instantiate it?

hovanleong commented on src/main/java/killua/command/AddCommand.java:
```diff
@@ -0,0 +1,42 @@
+package killua.command;
```
I really like that you split your files into many appropriate packages!

hovanleong commented on src/main/java/killua/parser/Parser.java:
```diff
@@ -0,0 +1,242 @@
+        String commandWord = parts[0];
+        String arguments = parts.length > 1 ? parts[1] : "";
+
+        switch (commandWord) {
```
Good job on obeying coding standards for your switch statement.

hovanleong commented on src/main/java/killua/task/Task.java:
```diff
@@ -0,0 +1,86 @@
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Good job following the variable naming conventions here!

hovanleong reviewed (commented):
LGTM. Just a very small handful of nits to fix. Overall really really great code, coding standards and naming conventions were very largely obeyed. Great job!

huangjiaxi1111 committed:
Make Task abstract

Make Task abstract can prevent it from being instantiated

yizhong187 commented on src/main/java/killua/task/Deadline.java:
```diff
@@ -0,0 +1,96 @@
+     * @return A string representation of the deadline for saving.
+     */
+    @Override
+    public String toSave() {
```
Should the `toSave` method name be renamed into something more descriptive? Perhaps `getSaveString` or `toSaveString`? I noticed this name in several other places too.

yizhong187 commented on src/main/java/killua/ui/Ui.java:
```diff
@@ -0,0 +1,209 @@
+        System.out.println("  deadline <description> /by <yyyy-mm-dd> OR ");
+        System.out.println("  deadline <description> /by <yyyy-mm-dd hh:mm> - Add a new deadline task");
+        System.out.println("  event <description> /from <yyyy-mm-dd> /to <yyyy-mm-dd> OR ");
+        System.out.println("  event <description> /from <yyyy-mm-dd hh:mm> /to <yyyy-mm-dd hh:mm> - Add a new event task");
```
I didn't like how long this line was. Perhaps use some line wrapping?

yizhong187 commented on src/main/java/killua/parser/Parser.java:
```diff
@@ -0,0 +1,242 @@
+                LocalDate toDate = LocalDate.parse(dateTimeStringTo, DATE_FORMATTER);
+                return new Event(description, fromDate, toDate);
+            } catch (DateTimeParseException e2) {
+                throw new IllegalArgumentException("Invalid date format: from '" + dateTimeStringFrom + "' to '" + dateTimeStringTo + "'");
```
I find that this line is too long and I noticed the same issue in several other places too. Perhaps you can wrap some parts to the next line?

yizhong187 commented on src/main/java/killua/parser/Parser.java:
```diff
@@ -0,0 +1,242 @@
+
+        Task task;
+
+        if (taskType == 'T') {
```
Perhaps you can use a switch statement here to avoid the many `if` `else` clauses.

yizhong187 commented on src/main/java/killua/parser/Parser.java:
```diff
@@ -0,0 +1,242 @@
+ */
+public class Parser {
+
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
I like how you have all your constant names in **SCREAMING_SNAKE_CASE**!

yizhong187 commented on src/main/java/killua/ui/Ui.java:
```diff
@@ -0,0 +1,209 @@
+    /**
+     * Prints a separator line for formatting purposes.
+     */
+    public void printLine() {
```
I like how you have a `printLine` method to abstract out the printing of a line and makes your code cleaning. Should you do that for general printint too? So you would not have to keep using `System.out.println`.

yizhong187 commented on src/main/java/killua/Killua.java:
```diff
@@ -0,0 +1,67 @@
+     * @param args Command line arguments.
+     */
+    public static void main(String[] args) {
+        new Killua("./data/tasks.txt").run();
```
Should you declare your `"./data/tasks.txt"` as a `FILE_PATH` constant?

yizhong187 reviewed (commented):
**LGTM**! I felt that your code abided by the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html) for the most part. There are only minor details such as line length that you may have missed out, and fixing them would make your code even more perfect. 🙌

> **Line length should be no longer than 120 chars.** Try to keep line length shorter than 110 characters (soft limit). But it is OK to exceed the limit slightly (hard limit: 120 chars). If the line exceeds the limit, use line wrapping at appropriate places of the line.

huangjiaxi1111 committed:
Fix coding style

Change toSave to toSaveAsString for clearity
Shorten lines to be less than 110 character

huangjiaxi1111 committed:
Add checkstyle

Fix import styling issues using checkstyle

huangjiaxi1111 committed:
Add GUI using javaFX

huangjiaxi1111 committed:
Merge branch 'branch-A-CheckStyle'

huangjiaxi1111 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI using javaFX

huangjiaxi1111 committed:
Fix bug in Parser

Parsing command from user input uses the same formater as parsing from history

Change formater for history to "MMM d yyyy"

huangjiaxi1111 committed:
Add assertions

Main.java, MainWindow.java, and DialogBox.java do not have assertion handling.

Add assertions in these three classes.

This is to ensure that required components are loaded and instances are initialized successfully.

huangjiaxi1111 committed:
Merge pull request #1 from huangjiaxi1111/branch-A-Assertions

Add assertions

huangjiaxi1111 committed:
Merge branch 'master' of https://github.com/huangjiaxi1111/ip

* 'master' of https://github.com/huangjiaxi1111/ip:
  Add assertions

huangjiaxi1111 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions

huangjiaxi1111 committed:
Merge remote-tracking branch 'origin/master' into branch-A-Streams

huangjiaxi1111 committed:
Fix code quality issues

Long string not initialised at the beginning, some comments are missing, and duplicated code in Parser

Fix long string issues, add essential comments, and refactor methods in Parser

This is to ensure better code quality

huangjiaxi1111 committed:
Merge pull request #2 from huangjiaxi1111/branch-A-CodeQuality

Branch a code quality

huangjiaxi1111 committed:
Merge branch 'master' of https://github.com/huangjiaxi1111/ip

* 'master' of https://github.com/huangjiaxi1111/ip:
  Fix code quality issues

huangjiaxi1111 committed:
Merge remote-tracking branch 'origin/master' into branch-A-Streams

huangjiaxi1111 committed:
Add Stream

Implement Stream on getTaskOnDate and getTaskString.

Make the code cleaner.

huangjiaxi1111 committed:
Merge pull request #3 from huangjiaxi1111/branch-A-Streams

Branch a streams

huangjiaxi1111 committed:
Merge branch 'master' of https://github.com/huangjiaxi1111/ip

* 'master' of https://github.com/huangjiaxi1111/ip:
  Add Stream

huangjiaxi1111 committed:
Add CI

huangjiaxi1111 committed:
Add detect duplicate

Same task can be added to task list multiple times.

Add detect duplicate in AddCommand to prevent adding duplicated task.

huangjiaxi1111 committed:
Merge branch 'branch-C-DetectDuplicates'

huangjiaxi1111 committed:
Tweak Gui

huangjiaxi1111 committed:
Merge branch 'branch-A-BetterGui'

huangjiaxi1111 committed:
Add user guide

huangjiaxi1111 committed:
Merge branch 'branch-A-UserGuide'

huangjiaxi1111 committed:
Fix User Guide

huangjiaxi1111 committed:
Add personality

Previous responses from Killua chatbot is rather boring.

Change the tone of Killua to be more impatient and sarcastic.

huangjiaxi1111 committed:
Add more personality

Killua's reponses to invalid inputs are boring.

Change Killua's tone to be sarcastic.

huangjiaxi1111 committed:
Update screenshot

Screenshot of the GUI is outdated after adding personality.

Update the screenshot with current GUI.

huangjiaxi1111 committed:
Update user guide

Commands lack more detailed descriptions.

Add more descriptions to each command.

huangjiaxi1111 committed:
Handle image not found

User image and Killua image may not be found in the specified folder.

Add handling of such case.

huangjiaxi1111 committed:
Fix bug: invalid index for mark/unmark

When key in "mark -1", killua doesn't handle the exception.

Add exception handling to catch IndexOutOfBoundsException. Also, fix
the issue where user mark/unmark marked/unmarked task.
